---
title: "Data Visualization with ggplot2 workbook"
author: "Junqi Lu"

output:
  html_document: 
    
    toc: true 
    toc_float: 
      collapsed: false 
      smooth_scroll: false 
      
    number_sections: true 
    
    theme: paper 
    
    output: 
      fig_width: 10 
      fig_height: 8 
      fig_caption: true 
      dev: svg 
    
    #Dataframe printing
    df_print: paged 
    
    code_folding: show 
  
  pdf_document: default
---

Batch save plots (Advanced; skip for now)
```{r}
knitr::opts_chunk$set(
  echo = TRUE,
  results = 'show',
  fig.path = "OutputPlots/",
  dev = c('svg', 'png'),
  
  
  fig.width = 10,
  #In the unit of inches
  fig.height = 8,
  #In the unit of inches
  unit = "in"
) 
```

# Load data
```{r}
library('scRNAseq')
scRNAData <- scRNAseq #Because ggplot2 builder needs the dataset to be in the Environment. Will cover later about ggplot2 builder
```

## Preview data structure 
```{r}
library('dplyr')
#Check on the shape of the dataset
glimpse(scRNAData)
```

```{r}
#Check the first several rows of the dataset
head(scRNAData)
```

It's always a good practice to check your data prior to making plots.

# ggplot2's layering logic
## Different connector
ggplot2 uses __"+"__ rather than __"%>%"__ to connect each command lines, or each layers. The different connector indicates a different logic. 

* In pipelines, commands or operations are run like water flowing inside a pipe. 
* In ggplot2, each layer of a plot is put together like staging in a theater. 

## Different layers of ggplots
Have the right order of which geom object should cover which 
```{r}
library('ggplot2')
```





## Aesthetics inheritance 
Remember the order of geom objects matter if you have multiple. If geom_boxplot() is after geom_point(), the boxplot will cover up lots of the points. 

Filter the original dataset
```{r}
three_organ_df <-
  scRNAData[scRNAData$transcript_length < 7500 &
              scRNAData$organ != 'other', ] #Don't forget the comma inside the brackets! 
#You can subset the dataset in the way you prefer. How would you do this by dplyr?

scRNAData %>% 
  filter(transcript_length < 7500,
         organ != 'other')
```

If a geom object does not have its own aes(), then it will refer to the global aes().
```{r}
ggplot(three_organ_df) +
  aes(x = organ,
      y = transcript_length,
      color = "red") + #pseudo-color; use scale_color_manual( breaks = NULL , values = "red") for real-coloring all geom into red
  
  geom_boxplot() + #Remember the order of geom objects matter if you have multiple. If geom_boxplot() is after geom_point(), the boxplot will cover up lots of the points
  geom_point() +
  
  labs(
    title = "Transcript length and GC content for genes in each organ",
    x = "Organ names",
    y = "Transcript length (bp)",
    color = 'GC content'
  ) +
  theme_minimal()
```
A geom object can have its own aes() for more specification.

Think about where you should move that aes(color='red') into and try yourself. 
Below I have the red boxplots with black point plots. 
```{r}
ggplot(three_organ_df, aes(x = organ,
      y = transcript_length)) +
  geom_boxplot(aes(color = 'red')) + 
  geom_point() +
  labs(
    title = "Transcript length and GC content for genes in each organ",
    x = "Organ names",
    y = "Transcript length (bp)",
    color = 'GC content'
  ) +
  theme_minimal()
```

Or I can have black boxplots and red point plots. 
```{r}
ggplot(three_organ_df, aes(x = organ,
      y = transcript_length)) +
  geom_boxplot() + 
  geom_point(colour = "red") +
  labs(
    title = "Transcript length and GC content for genes in each organ",
    x = "Organ names",
    y = "Transcript length (bp)",
    color = 'GC content'
  ) +
  theme_minimal()
```

(Back to slide.)

# ggplot2 builder
Use __ggplot2 builder__ for general data exploration, output codes, and provide more edits on the code. 

* Installation: install.packages('esquisse') and _restart your RStudio_. Then you can find the ggplot2 builder under the Addin dropdown menu

Idea: let's compare how cell 1's transcriptome is different from cell 2's. What plot might work the best for this job? 

## Raw output
```{r}

library(ggplot2)

ggplot(three_organ_df) +
 aes(x = cell_1, y = cell_2) +
 geom_point(shape = "circle", size = 1.5, colour = "#112446") +
 theme_minimal()

```

## Edited with more details 
* Add a geom_abline to split 
* facet_wrap for vars(organ)
* labels & theme
```{r}
ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  facet_wrap(vars(organ), labeller = label_both) +
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  theme_minimal() #I personally like to move the theme section to the end
```

It seems like the the differene in transcriptomes of cell 1 and 2 are mainly in the genes expressed in the heart and kidney the most.  

(Back to slide.)

# More power to ggplot2 
## viridis
__install.packages("viridis")__ and __library(viridis)__
```{r}
library("viridis")
ggplot(three_organ_df) + #Filtering the dataset on the run
  aes(x = organ,
      y = transcript_length,
      color = gene_percent_gc_content) +
  geom_point() +
  facet_wrap( ~ strand) +
  labs(
    title = "Transcript length and GC content for genes in each organ",
    x = "Organ names",
    y = "Transcript length (bp)",
    color = 'GC content'
  ) +
  theme_minimal()+
  scale_color_viridis(direction = -1)
```




## ggsignif
It caluclates p-values for you on-the-fly. 
__install.packages("ggsignif")__ and __library(ggsignif)__
```{r}
library("ggsignif")

ggplot(three_organ_df) +
  aes(x = organ,
      y = transcript_length) +
  
  geom_boxplot() + #Remember the order of geom objects matter if you have multiple. If geom_boxplot() is after geom_point(), the boxplot will cover up lots of the points
  geom_signif(comparisons = list(c("brain", "heart"))) +
  geom_signif(
    comparisons = list(c("heart", "other")),
    map_signif_level = TRUE,
    #This converts the p-value to significant annotations
    vjust = 0.1 #This move up the annotation up a bit from the bar
  ) +
  geom_signif(
    comparisons = list(c("brain", "other")),
    map_signif_level = TRUE,
    #This converts the p-value to significant annotations
    vjust = 0.1,
    #This moves up the annotation up a bit from the bar
    
    y_position = 8100,
    #This moves up the significant bar to a specific y level
    annotation = 'Significant!',
    #This overwrites the text for the annotation
  ) +
  
  
  labs(title = "Transcript length and GC content for genes in each organ",
       x = "Organ names",
       y = "Transcript length (bp)") +
  theme_minimal()
```



## ggExtra
__install.packages("ggExtra")__ and __library(ggExtra)__

It comes with an addin called ggMarginal() for interactive data exploration. 

You first create the major plot by ggplot2.
```{r}
library("ggExtra")

ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  theme_minimal() #I personally like to move the theme section to the end
```

Then you highlight the ggplot2 code chunk and use the addin, adjust the extra plots around the major plot, and output the code. 
```{r}
p1 <- ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  theme_minimal() #I personally like to move the theme section to the end

ggExtra::ggMarginal(
  p = p1,
  type = 'boxplot', #You can use other type of marginal plots here
  margins = 'both',
  size = 3.5,
  colour = 'black',
  fill = 'gray'
)
```

(Back to slide.)

# Save a favorite style to use later
Sometimes your PI might want to your graphs to look in a specific style that he or she likes. Instead of coping and pasting the theme codes every time for every graph, you can save it in as a list and call it whenever you need it. 

Original plot:
```{r}
ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  facet_wrap(vars(organ), labeller = label_both) +
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  theme_minimal() 
```

What my PI likes:
```{r}
ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  facet_wrap(vars(organ), labeller = label_both) +
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  theme_linedraw() +
  
  
  #Below is the style that my PI specifically likes
  theme(
    #For all the fonts, the default font is "Arial" and the all the sizes are in the units of pt
    
    plot.title = element_text(size = 12, face = "bold"),
    plot.subtitle = element_text(size = 10, face = "bold", color = "black"),
    panel.border = element_rect(size = 1.5, color = "black"),
    
    axis.line = element_line(colour = '1.5', size = 1.5),
    #Jennifer likes the axis line to be of 1.5 pt thick
    axis.title = element_text(size = 10, face = "bold"),
    axis.text = element_text(size = 10, face = "bold"),
    
    strip.text.x = element_text(size = 10, face = "bold", color = "black"),
    strip.text.y = element_text(
      size = 10,
      face = "bold",
      color = "black",
      angle = 0
    ),
    strip.background = element_blank(),
    
    legend.title = element_text(size = 10, face = "bold"),
    legend.text = element_text(size = 9, face = "bold"),
    legend.key.width = unit(10, "pt") #Control the size of the legend unit. Decrease this if you have a very wide legend, which is unnecessary
  )
 
```

Save the specific styling in a list and call on it. 
```{r}
MyStyle = list(
  theme_linedraw() +
    
    
    #Below is the style that my PI specifically likes
    theme(
      #For all the fonts, the default font is "Arial" and the all the sizes are in the units of pt
      
      plot.title = element_text(size = 12, face = "bold"),
      plot.subtitle = element_text(size = 10, face = "bold", color = "black"),
      panel.border = element_rect(size = 1.5, color = "black"),
      
      axis.line = element_line(colour = '1.5', size = 1.5),
      #Jennifer likes the axis line to be of 1.5 pt thick
      axis.title = element_text(size = 10, face = "bold"),
      axis.text = element_text(size = 10, face = "bold"),
      
      strip.text.x = element_text(size = 10, face = "bold", color = "black"),
      strip.text.y = element_text(
        size = 10,
        face = "bold",
        color = "black",
        angle = 0
      ),
      strip.background = element_blank(),
      
      legend.title = element_text(size = 10, face = "bold"),
      legend.text = element_text(size = 9, face = "bold"),
      legend.key.width = unit(10, "pt") #Control the size of the legend unit. Decrease this if you have a very wide legend, which is unnecessary
    )
  
)


ggplot(three_organ_df) +
  aes(x = cell_2, y = cell_1) +
  
  geom_point(
    shape = "circle",
    size = 1.5,
    colour = 'blue',
    alpha = 0.3
  ) +
  geom_abline(intercept = 0,
              color = 'green') + #As a reference of how different
  
  
  facet_wrap(vars(organ), labeller = label_both) +
  labs(title = "Cell 1 compared to Cell 2",
       x = "Cell 1 transcriptome",
       y = "Cell 2 transcriptome") +
  MyStyle
```

(Back to slide.)